### PR TITLE
[GH-1000] Select character functionality

### LINKED ISSUE #1000: [FEATURE] Select character
## 🚀 Feature
- Add a button in character list cards to select a character and store it in the game session. It should also redirect the player to the main screen.
- The stored character's model must be the one to appear in the main screen scene.
- When a game starts, we need to use the selected character instead of a random one.
- While this doesn't fit inside the scope of this issue, currently there isn't a way to create a user, and the selected character in the backend needs a user to be assigned to, so a user create when the application is started may be added for the character selection functionality to work.

### PR DESCRIPTION
**Pull request by tkz00**
Closes #1000 

Works in conjunction with [this backend PR](https://github.com/lambdaclass/game_backend/pull/61).

Waiting on #1237 

## Motivation

Users must be able to select the character they want to play with.

#### It's also needed to modify the Utils script to point towards the localhost server before getting to the main screen (where you can choose the server manually). This is done by changing the `MakeHTTPUrl` function of the class to:

```
    public static string MakeHTTPUrl(string path)
    {
        return "http://" + "localhost" + ":4000" + path;
    }
```

There is a simplify user creation flow explanation:
UserCreation drawio

There is a simplify character Selection flow explanation:
Captura de pantalla 2023-12-13 a la(s) 13 31 26

- To test user creation on title screen:
  - You must not have an user in the database with your `device_client_id`.
  - Start the game and after you click the ENTER button in title screen scene a loading spinner should appear, then you will be sent to main screen and a user with your `device_client_id` should have been created in the database.
- To test character selection:
  - Go to character list scene, choose one and select it in the character info scene.
  - You will be taken to the main scene and the character model to appear in the middle of the screen should be the one you chose.
  - Also the `selected_character` column in your database user should be the character chosen.
  - When you play a game the character you are assigned should be the one you have selected.

## Test additions / changes

1. Enable H4ck adding "H4ck" to the `availableCharacterNames` list in `CharactersManager`
2. In the scriptableObject of H4ck add the Muflus prefab to Prefab field and Muflus UImodel to the UImodel field (This is because we dont have th h4ck model anymore)
Captura de pantalla 2023-12-21 a la(s) 12 25 00

3. Run the backend and the client
4. Go to characterList select Muflus Or H4ck you should see the name of the selected in the LobbyConnection GO in the field `selectedCharacterName`
5. Play a game you should have the selected
6.  Kill the client. Start again you should have the last selected character as you starter character
7. Repeat

## Checklist
- [X] I have tested the changes locally.
- [x] I self-reviewed the changes on GitHub, line by line.
- [ ] Tests have been added/updated.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [x] I have tested the changes in another devices.
  - [ ] Tested in iOS.
  - [x] Tested in Android.

### CODE REVIEW

**Review by agustinesco — CHANGES_REQUESTED** [review]
image
image

When i try to enter the game the loading spinner stay there forever, i checked the user db has no user registered, is there something else i need to do? also i tested this with the corresponding back repo pr

File: client/Assets/Scripts/UI/UIModelManager.cs [hunk 1]
```diff
@@ -27,6 +28,11 @@ public void SetModel(string characterName)
         StartCoroutine(AnimateCharacter(modelClone));
     }
 
+    // public void SetupList(List<string> characters)
+    // {
+    //     enabledCharacters = characters;
+    // }
+
```
Comment (ncontinanza) [inline client/Assets/Scripts/UI/UIModelManager.cs:35]: Should we remove this commented code?

File: client/Assets/Scripts/Utils/ServerUtils.cs [hunk 2]
```diff
@@ -4,9 +4,18 @@
...
 using UnityEngine.Networking;
+using System.Text;
 
 public static class ServerUtils
 {
+    class AcceptAllCertificates : CertificateHandler
+    {
+        protected override bool ValidateCertificate(byte[] certificateData)
+        {
+            return true;
```
Comment (ncontinanza) [inline client/Assets/Scripts/Utils/ServerUtils.cs:15]: What's the use of this?

File: client/Assets/Scripts/Utils/ServerUtils.cs [hunk 3]
```diff
@@ -60,6 +69,7 @@ Action<string> errorCallback
                     UserCharacterResponse response = JsonUtility.FromJson<UserCharacterResponse>(
                         webRequest.downloadHandler.text
                     );
+                    Debug.Log(webRequest.downloadHandler.text);
```
Comment (ncontinanza) [inline client/Assets/Scripts/Utils/ServerUtils.cs:72]:
```suggestion

```

**Review by ncontinanza — CHANGES_REQUESTED** [review]

File: client/Assets/Scripts/CharactersManager.cs [hunk 4]
```diff
@@ -1,23 +1,70 @@
...
+    {
+        return this.GoToCharacter;
+    }
+
+    public List<string> GetAvailableCharactersNames()
+    {
+        return availableCharacterNames;
+    }
+
+    public bool IsAvaibleCharacter(CoMCharacter character)
```
[the remark this answers is not in the document]
Comment (BertovDev) [reply client/Assets/Scripts/CharactersManager.cs]: Right!

File: client/Assets/Scripts/LobbyConnection.cs [hunk 5]
```diff
@@ -397,6 +399,8 @@ private void OnWebSocketMessage(byte[] data)
                     break;
 
                 case LobbyEventType.NotifyPlayerAmount:
+                    this.playerCount = (int)lobbyEvent.AmountOfPlayers;
```
[the remark this answers is not in the document]
Comment (BertovDev) [reply client/Assets/Scripts/LobbyConnection.cs]: This is a merge error, good catch

File: client/Assets/Scripts/UI/UIModelManager.cs (hunk 1 above)
Comment (BertovDev) [reply client/Assets/Scripts/UI/UIModelManager.cs:35]: Yess

File: client/Assets/Scripts/Utils/ServerUtils.cs (hunk 2 above)
Comment (BertovDev) [reply client/Assets/Scripts/Utils/ServerUtils.cs:15]: This is to implement a custom certificate validation scheme. But i believe we dont want it.

**Review by ncontinanza — APPROVED** [review]

**Review by AngieDutra — APPROVED** [review]
This PR has this [bug](https://github.com/lambdaclass/curse_of_mirra/issues/1302) if more than one character is available. Will do a PR fixing this asap

**Pull request merged**